File: AndroidApp/app/src/main/java/com/ep133/sampletool/webview/MIDIBridge.kt
package com.ep133.sampletool.webview

import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.ep133.sampletool.midi.MIDIPort
import org.json.JSONArray
import org.json.JSONObject

/**
 * JavaScript interface exposed to the WebView as `window.EP133Bridge`.
 *
 * The shared polyfill detects this object and routes MIDI calls through it:
 *   - getMidiDevices() → returns JSON string of available MIDI ports
 *   - sendMidi(portId, dataJson) → sends MIDI bytes to a device
 *
 * Incoming MIDI is pushed to JS via [forwardMIDIToJS].
 */
class MIDIBridge(
    private val midiManager: MIDIPort,
    private val webView: WebView
) {

    @JavascriptInterface
    fun getMidiDevices(): String {
        val devices = midiManager.getUSBDevices()

        val inputs = JSONArray().apply {
            devices.inputs.forEach { d ->
                put(JSONObject().apply {
                    put("id", d.id)
                    put("name", d.name)
                })
            }
        }

        val outputs = JSONArray().apply {
            devices.outputs.forEach { d ->
                put(JSONObject().apply {
                    put("id", d.id)
                    put("name", d.name)
                })
            }
        }

        return JSONObject().apply {
            put("inputs", inputs)
            put("outputs", outputs)
        }.toString()
    }

    @JavascriptInterface
    fun sendMidi(portId: String, dataJson: String) {
        val arr = JSONArray(dataJson)
        val bytes = ByteArray(arr.length()) { i ->
            arr.getInt(i).toByte()
        }
        midiManager.sendMidi(portId, bytes)
    }

    /**
     * Pushes incoming MIDI data from a device to the WebView's JavaScript.
     * Called from [MIDIManager.onMidiReceived].
     */
    fun forwardMIDIToJS(portId: String, data: ByteArray) {
        val dataArray = data.joinToString(",") { (it.toInt() and 0xFF).toString() }
        val escapedPortId = portId.replace("'", "\\'")
        val js = "window.__ep133_onMidiIn('$escapedPortId', [$dataArray])"
        webView.post { webView.evaluateJavascript(js, null) }
    }

    /**
     * Notifies the web app that MIDI devices have changed.
     * Triggers the polyfill's statechange handler and also dispatches a
     * custom event the web app can listen for.
     */
    fun notifyDevicesChanged() {
        val js = """
            if (window.__ep133_onDevicesChanged) {
                window.__ep133_onDevicesChanged();
            }
        """.trimIndent()
        webView.post { webView.evaluateJavascript(js, null) }
    }
}

File: iOSApp/EP133SampleTool/WebView/MIDIBridge.swift
import WebKit
import CoreMIDI

/// Bridges JavaScript MIDI calls from the WebView to CoreMIDI.
///
/// The polyfill sends messages via:
///   window.webkit.messageHandlers.midibridge.postMessage({action, callbackId, portId, data})
///
/// This handler processes them and resolves callbacks via:
///   window.__ep133_resolveCallback(callbackId, jsonString)
///
/// Incoming MIDI is pushed to JS via:
///   window.__ep133_onMidiIn(portId, [bytes])
final class MIDIBridge: NSObject, WKScriptMessageHandler {

    weak var webView: WKWebView?
    private let midiManager = MIDIManager()

    override init() {
        super.init()
        midiManager.onMIDIReceived = { [weak self] portId, data in
            self?.forwardMIDIToJS(portId: portId, data: data)
        }
        midiManager.onDevicesChanged = { [weak self] in
            // Notify the web app that devices changed so it can re-enumerate
            self?.evaluateJS("if (navigator.requestMIDIAccess) { console.log('[EP133] MIDI devices changed'); }")
        }
        midiManager.setup()
    }

    // MARK: - WKScriptMessageHandler

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? [String: Any],
              let action = body["action"] as? String else {
            return
        }

        switch action {
        case "getMidiDevices":
            handleGetMidiDevices(callbackId: body["callbackId"] as? String)

        case "sendMidi":
            handleSendMidi(
                portId: body["portId"] as? String,
                data: body["data"] as? [Int]
            )

        default:
            print("[EP133] Unknown MIDI bridge action: \(action)")
        }
    }

    // MARK: - Action Handlers

    private func handleGetMidiDevices(callbackId: String?) {
        let devices = midiManager.getUSBDevices()

        let inputsJSON = devices.inputs.map { d in
            "{\"id\":\"\(escapeJS(d.id))\",\"name\":\"\(escapeJS(d.name))\"}"
        }.joined(separator: ",")

        let outputsJSON = devices.outputs.map { d in
            "{\"id\":\"\(escapeJS(d.id))\",\"name\":\"\(escapeJS(d.name))\"}"
        }.joined(separator: ",")

        let json = "{\"inputs\":[\(inputsJSON)],\"outputs\":[\(outputsJSON)]}"

        if let cbId = callbackId {
            resolveCallback(callbackId: cbId, json: json)
        }
    }

    private func handleSendMidi(portId: String?, data: [Int]?) {
        guard let portId = portId, let data = data else { return }
        let bytes = data.map { UInt8(clamping: $0) }
        midiManager.sendMIDI(to: portId, data: bytes)
    }

    // MARK: - JS Communication

    private func forwardMIDIToJS(portId: String, data: [UInt8]) {
        let dataArray = data.map { String($0) }.joined(separator: ",")
        let js = "window.__ep133_onMidiIn('\(escapeJS(portId))', [\(dataArray)])"
        evaluateJS(js)
    }

    private func resolveCallback(callbackId: String, json: String) {
        let escapedJSON = json.replacingOccurrences(of: "'", with: "\\'")
        let js = "window.__ep133_resolveCallback('\(escapeJS(callbackId))', '\(escapedJSON)')"
        evaluateJS(js)
    }

    private func evaluateJS(_ script: String) {
        DispatchQueue.main.async { [weak self] in
            self?.webView?.evaluateJavaScript(script)
        }
    }

    private func escapeJS(_ str: String) -> String {
        str.replacingOccurrences(of: "\\", with: "\\\\")
           .replacingOccurrences(of: "'", with: "\\'")
           .replacingOccurrences(of: "\"", with: "\\\"")
           .replacingOccurrences(of: "\n", with: "\\n")
    }
}
